File: 프로그래머스/lv2/김원래/연속 부분 수열 합의 개수.cpp
#include <string>
#include <vector>
#include <set>

using namespace std;

set<int> res;
int n;


int solution(vector<int> elements) {
    int answer = 0;
    
    int sum[2001];
    
    n = elements.size();
    
    for(int cnt = 0 ; cnt < n ; ++cnt) {
        elements.push_back(elements[cnt]);
    }
    
    sum[0] = elements[0];
    
    for(int cnt = 1; cnt < elements.size() ; ++cnt) {
        sum[cnt] = sum[cnt - 1] + elements[cnt];
    }
    

    for(int range = 0 ; range < n ; ++range)
    {
        for(int start = 0 ; start < elements.size() - range; ++start)
        {
            int curSum = 0;
            
            if(start == 0) {
                curSum = sum[start + range];
            }
            else {
                curSum = sum[start + range] - sum[start - 1];
            }
            
            if(res.find(curSum) == res.end()) res.insert(curSum);
        }
    }
    answer = res.size();
    
    return answer;
}



/*

이전에 풀었던 dp문제와 비슷해보인다.
오히려 더 쉬운 문제라고 생각이 듦

dp[n] = 0부터 n까지의 합

*/

File: 프로그래머스/lv2/장정윤/연속 부분 수열 합의 개수.py
from itertools import product
def solution(elements):
    answer = 0
    n = len(elements)
    # n개 연속 합치고, 결과 값 중복은 제거
    
    # 순환 수열 리스트 2배
    elements = elements * 2
    unique_values = set()
    
    for end in range(1, n + 1):
        for start in range(n):
            #start부터 end idx 원소 더함
            sum_value = sum(elements[start:start+end])
            unique_values.add(sum_value)
        
    return len(unique_values)